### DISCUSSION TITLE
auto highlight doesn't seem to work for MVT layers with more than one instance of a shape that matches the uniquePropertyId

### DISCUSSION THREAD

**Discussion by landtechjoe** [visgl/deck.gl#8950] [Q&A] [answered]
I am guessing this may be by design as the property is called `uniquePropertyId` but sometimes, due to the nature of MVT features get broken up in such a way that within a tile there may be 2 items drawn in one tile that have the same id. A V-shaped piece of land for instance - if the top part of the V was rendered on one tile and the rest on the tile below. There would be 2 instances of that id present in the top tile for each side of the top of the V.

From what I can see the code behind the `autoHighlight` uses `indexOf` to find the first matching element and so that gets highlighted but other ones do not. 

I have hacked together an example that roughly approximates this behaviour here https://codepen.io/Joe-Coulson/pen/VwOrRbp If you hover over the blue in the center of the image you'll see some highlighted content. You'll also see that there is a chunk missing from the center of the polygon that is not highlighted. This is caused by the presence (I believe) of the point within the same tile that has a matching `uniquePropertyId` so that is the only item that gets rendered.

Is there a simple way around this that anyone knows of or am I am going to make another solution? I know I can do a solution using `getFillColor` and setting a property on hover to update things however we have some complex mapping needs and that approach degrades performance too much.

I've attached a "beautiful" drawing to illustrate. Outer circle covers missing area inner circle is the culprit. Also hovering over the non-highlighted area still highlights the other items with matching `uniquePropertyId` so they certainly match

Any suggestions on how to fix this?

![Screenshot 2024-06-14 at 12 19 17](https://github.com/visgl/deck.gl/assets/97101845/6afaa4c7-38db-41c8-8db0-815a449d2695)

**Comment by Pessimistress** [comment, marked as answer]
Did you generate the MVT tiles yourself? The regions with identical ids should be represented by a single MultiPolygon feature, instead of several Polygon features. Every geometry originally from the same feature gets hightlighted together.

**Reply by landtechjoe** [reply]
Thanks for your answer. Your knowledge got me right to where I needed to be. We are generating the tiles ourselves but the tiles were being generated (in some instances) from discrete rows in our database so we weren't grouping on what was set as the `uniqueIdProperty`. Updating our query to group things correctly then started to return these items as a single MultiPolygon.

**Reply by smcallis** [reply]
I'm running into this myself.  What about features that contain geometry of different dimension?   E.g. a polygon and a point.  E.g. this:

![image](https://github.com/user-attachments/assets/8e80745b-6929-4085-9060-3d136c474a62)

The only difference between the highlighted area on the left and the empty area on the right is that the tile on the right has a single extra point for the same feature (you can't see it).
